packages/amplication-server/src/core/UsageInsights/usageInsights.service.ts

```diff
@@ -0,0 +1,230 @@
+import { Prisma, PrismaService } from "../../prisma";
+import {
+  BaseUsageInsightsArgs,
+  BlockChangesArgs,
+  ParsedQueryRowResult,
+  QueryRawResult,
+} from "./types";
+import {
+  UsageInsightsResult,
+  EvaluationInsights,
+  UsageInsights,
+} from "./dtos/UsageInsights.object";
+import { EnumBlockType } from "../../enums/EnumBlockType";
+
+@Injectable()
+export class UsageInsightsService {
+  constructor(
+    private readonly logger: AmplicationLogger,
+    private readonly prisma: PrismaService
+  ) {}
+
+  async countLinesOfCode({
+    projectIds,
+    startDate,
+    endDate,
+  }: BaseUsageInsightsArgs): Promise<number> {
+    const aggregatedLoc = await this.prisma.build.aggregate({
+      where: {
+        linesOfCode: {
+          not: null,
+        },
+        createdAt: {
+          gte: startDate,
+          lte: endDate,
+        },
+        resource: {
+          project: {
+            id: {
+              in: projectIds,
+            },
+          },
+        },
+      },
+      _sum: {
+        linesOfCode: true,
+      },
+    });
+
+    return aggregatedLoc._sum.linesOfCode ?? 0;
+  }
+
+  async countBuilds({
+    startDate,
+    endDate,
+    projectIds,
+    timeGroup,
+  }: BaseUsageInsightsArgs): Promise<UsageInsights> {
+    const results: QueryRawResult[] = await this.prisma.$queryRaw`
+    SELECT DATE_PART('year', b."createdAt") as year, DATE_PART('month', b."createdAt") as month, DATE_PART(${timeGroup}, b."createdAt") as time_group, COUNT(b.*) AS count
+    FROM "Build" b
+    JOIN "Resource" r ON b."resourceId" = r."id"
+    WHERE b."createdAt" >= ${startDate}
+    AND b."createdAt" <= ${endDate}
+    AND r."projectId" IN (${Prisma.join(projectIds)})
+    GROUP BY year, month, time_group
+    ORDER BY year, month, time_group;
+  `;
+
+    return {
+      results: Object.values(this.translateToAnalyticsResults(results)),
+    };
+  }
+
+  async countEntityChanges({
+    projectIds,
+    startDate,
+    endDate,
+    timeGroup,
+  }: BaseUsageInsightsArgs): Promise<UsageInsights> {
+    const results: QueryRawResult[] = await this.prisma.$queryRaw`
+    SELECT DATE_PART('year', ev."updatedAt") as year, DATE_PART('month', ev."updatedAt") as month, DATE_PART(${timeGroup}, ev."updatedAt") as time_group, COUNT(ev.*) AS count
+    FROM "EntityVersion" ev
+    JOIN "Entity" e ON ev."entityId" = e."id"
+    JOIN "Resource" r ON e."resourceId" = r."id"
+    WHERE r."projectId" IN (${Prisma.join(projectIds)})
+    AND ev."updatedAt" >= ${startDate} AND ev."updatedAt" <= ${endDate}
+    GROUP BY year, month, time_group
+    ORDER BY year, month, time_group;
+  `;
+
+    return {
+      results: Object.values(this.translateToAnalyticsResults(results)),
+    };
+  }
+
+  async countBlockChanges({
+    projectIds,
+    startDate,
+    endDate,
+    blockType,
+    timeGroup,
+  }: BlockChangesArgs): Promise<UsageInsights> {
+    let results: QueryRawResult[];
+    switch (blockType) {
```

Comment:
It's not critical, but I still feel we can use a mapper from BlockType to string somehow and avoid duplication here.

Reply:
As I told you, it didn't work. I can show you what I did, maybe I missed something